torch/distributed/pipelining/schedules.py

```diff
@@ -2097,71 +2166,331 @@ def backward_stage_index(step):
+            self.pipeline_order,
+            self.pp_group_size,
+        )
 
-            for rank in range(self.pp_group_size):
-                timestamp = next_pointer[rank]
-                if timestamp >= len(actions[rank]):
+    def _get_zbv_schedule(
+        self,
+        num_model_chunks,
+        pipeline_parallel_size,
+        num_microbatches,
+    ):
+        pp_group_size = pipeline_parallel_size
+        mb_size = num_microbatches
+        num_local_stages = num_model_chunks
+        compute_schedules = {}
+        category_map = {
+            F: 0,
+            I: 1,
+            W: 2,
+        }
+        chunk_0 = 0
+        chunk_1 = 1
+        def get_compute_schedule(pipeline_parallel_size, num_microbatches):
+            num_chunks = 2
+            n_node = len(category_map) * num_chunks * pipeline_parallel_size * num_microbatches
+
+            def get_id(op, chunk, rank, microbatch_id):
+                category = category_map[op]
+                return (
+                    category * 2 * pipeline_parallel_size * num_microbatches
+                    + chunk * pipeline_parallel_size * num_microbatches
+                    + rank * num_microbatches
+                    + microbatch_id
+                )
+            ops_count = []
+
+            for i in range(pipeline_parallel_size):
+                temp = {}
+                temp[F] = [0] * 2
+                temp[I] = [0] * 2
+                temp[W] = [0] * 2
+                ops_count.append(temp)
+
+            fbw_mem = [3, -1, -2]
+            max_mem = 3 * (pipeline_parallel_size * 2)
+            end_time = [-1] * n_node
+            cur_time = [0] * pipeline_parallel_size
+            mem = [0] * pipeline_parallel_size
+            rank_bubble = [0] * pipeline_parallel_size
+            pending_w: List[Deque[Tuple[int, int, int]]] = [
+                deque() for _ in range(pipeline_parallel_size)
+            ]
+            schedule: Dict[int, List[Tuple[int, _ComputationType, int]]] = {}
+            for i in range(pipeline_parallel_size):
+                schedule[i] = []
+
+            def put_w(rank):
+                assert len(pending_w[rank]) > 0
+                _, chunk_, _ = pending_w[rank].popleft()
+                put(W, chunk_, rank)
+
+            def put(op, chunk, rank, assert_cnt=True):
+                """
+                Schedules an operation (forward, backward, or weight update) for a given chunk and rank.
+                op (_ComputationType): The type of operation (FORWARD, BACKWARD, WEIGHT).
+                chunk (int): The model chunk index (0 or 1).
+                rank (int): The pipeline stage index.
+                """
+                _cur_time = _no_bubble = cur_time[rank] + 1
+                category = category_map[op]
+                # Number of ops (F/B/W) with the same (action, chunk) on current rank
+                _op_count = ops_count[rank][op][chunk]
+                if chunk == chunk_1 and op in (F, I):
```

Comment:
maybe these 2 ifs would make for a nice helper function whose name would further clarify their purpose.  IIUC the idea here is to figure out what the earliest time this op/chunk can run on the current rank, given the time that its dependency is scheduled on another rank?

"get_earliest_time_based_on_dependency(op, chunk, rank)"